docs: Update Threat Model for Karpenter 1.0.0 (#6893)
aws/karpenter-provider-aws: website/content/en/docs/reference/threat-model.md

#### Karpenter can orchestrate the creation/deletion of IAM Instance Profiles it doesn’t own

**Threat:** Karpenter has permission to create/update/delete instance profiles as part of its controller permissions to ensure that it can auto-generate instance profiles when `EC2NodeClasses` are created. An actor who obtains control of the Karpenter pod’s IAM role may delete instance profiles not owned by Karpenter, causing workload disruption to other instances using the profile in the account.

**Mitigation:** Karpenter’s controller permissions enforce that it creates instance profiles with tags which provide ownership information. These tags include:

* `karpenter.sh/managed-by`
* `kubernetes.io/cluster/${CLUSTER_NAME}`
* `karpenter.k8s.aws/ec2nodeclass`
* `topology.kubernetes.io/region`

Source: https://karpenter.sh/docs/reference/cloudformation/#allowscopedinstanceprofilecreationactions

> These tags ensure that instance profiles created by Karpenter in the account are unique to that cluster. Karpenter’s controller permissions only allow it to act on instance profiles that contain these tags which match the cluster information.

Reviewer: Can we use this same wording for instance permissions that we list above -- seems like similar wording in both of these locations would be nice

#### Operator and developer have different permissions

> **Threat:** When installing Karpenter, cluster operators define an [IAM Role](https://karpenter.sh/docs/reference/cloudformation/#controller-authorization) for Karpenter to run under. Additionally, cluster operators must provide an [Instance IAM Role](https://karpenter.sh/docs/reference/cloudformation/#node-authorization) or Instance Profile for EC2 Instance launched by Karpenter. If the developer had permission to exec into the Karpenter pod, they would elevate their permissions to that of the IAM role for Karpenter.

Reviewer: Can we link to the instance profile doc?

> **Threat:** When installing Karpenter, cluster operators define an [IAM Role](https://karpenter.sh/docs/reference/cloudformation/#controller-authorization) for Karpenter to run under. Additionally, cluster operators must provide an [Instance IAM Role](https://karpenter.sh/docs/reference/cloudformation/#node-authorization) or Instance Profile for EC2 Instance launched by Karpenter. If the developer had permission to exec into the Karpenter pod, they would elevate their permissions to that of the IAM role for Karpenter.

Reviewer: ```suggestion
**Threat:** When installing Karpenter, cluster operators define an [IAM Role](https://karpenter.sh/docs/reference/cloudformation/#controller-authorization) for Karpenter to run under. Additionally, cluster operators must provide an [Instance IAM Role](https://karpenter.sh/docs/reference/cloudformation/#node-authorization) or an Instance Profile for EC2 Instances launched by Karpenter. If the developer has permission to exec into the Karpenter pod, they can escalate their permissions to that of the IAM node role for Karpenter.
```

> **Mitigation:** Karpenter runs in a separate namespace [(currently recommended to run under kube-system namespace)](https://karpenter.sh/docs/getting-started/getting-started-with-karpenter/#preventing-apiserver-request-throttling) , which enables cluster operators to restrict access using Kubernetes RBAC. 

Reviewer: ```suggestion
**Mitigation:** Karpenter runs in a separate namespace [(currently recommended to run under kube-system namespace)](https://karpenter.sh/docs/getting-started/getting-started-with-karpenter/#preventing-apiserver-request-throttling), which enables cluster operators to restrict access using Kubernetes RBAC. 
```

Can we do an audit of spacing throughout this doc? I see spaces randomly thrown in in a bunch of places throughout and I'm not sure if that's going to affect the formatting that we see when this is rendered in our docs.

#### Undesired Disruption Through NodePool Access

> Problem: If the user is able to modify the `NodePool` for a given budget, they can edit the budgets to change the resulting behavior. This could either terminate nodes when the user doesn’t want it, or block Karpenter to terminate and patch nodes when a user does. 

Reviewer: ```suggestion
**Threat**: If the user is able to modify the `NodePool` for a given budget, they can edit the budgets to change the resulting behavior. This could either terminate nodes when the user doesn’t want it, or block Karpenter to terminate and patch nodes when a user does. 
```

Can this have the `Threat/Mitigation` syntax that we've been using throughout the doc?

> Problem: If the user is able to modify the `NodePool` for a given budget, they can edit the budgets to change the resulting behavior. This could either terminate nodes when the user doesn’t want it, or block Karpenter to terminate and patch nodes when a user does. 

Reviewer: ```suggestion
Problem: If the user is able to modify the `NodePool`, they can edit the budgets to allow the disruption time windows to be more or less permissive than the user intended. This could either terminate nodes when the user doesn’t want it or block Karpenter to terminate and patch nodes when a user does. 
```

> Mitigation: Access to the `NodePool` is treated as a privileged permission. If an Application Owners had access to the `NodePool`, they could modify or delete everything about the existing nodes in the cluster. By default, only the Cluster Operator has access to this resource. If the Cluster Operator grants another user permission, this is the user’s responsibility. Karpenter also emits a metric for the current allowed disruption that users can monitor in case their `NodePools` are modified. Karpenter will also emit an event for the `NodePool` if disruptions are allowed. 

Reviewer: ```suggestion
**Mitigation**: Access to the `NodePool` is treated as a privileged permission. If an Application Owners had access to the `NodePool`, they could modify or delete everything about the existing nodes in the cluster. By default, only the Cluster Operator has access to this resource. If the Cluster Operator grants another user permission, this is the user’s responsibility. Karpenter also emits a metric for the current allowed disruption that users can monitor in case their `NodePools` are modified. Karpenter will also emit an event for the `NodePool` if disruptions are allowed. 
```

It's not bolded

> Mitigation: Access to the `NodePool` is treated as a privileged permission. If an Application Owners had access to the `NodePool`, they could modify or delete everything about the existing nodes in the cluster. By default, only the Cluster Operator has access to this resource. If the Cluster Operator grants another user permission, this is the user’s responsibility. Karpenter also emits a metric for the current allowed disruption that users can monitor in case their `NodePools` are modified. Karpenter will also emit an event for the `NodePool` if disruptions are allowed. 

Reviewer: ```suggestion
Mitigation: Access to the `NodePool` is treated as a privileged permission. If Application Owners had access to the `NodePool`, they could modify or delete any nodes on the cluster. By default, only the Cluster Operator should have access to this resource. Karpenter also emits metrics for the currently allowed disruptions that users can monitor to ensure their `NodePools` don't get modified.
```

Same comment here -- I'm not sure what the statement about granting other users access means. Do we even need this statement?

#### Launching EC2 instances with IAM roles not intended for Karpenter nodes

> **Threat:** Many IAM roles in an AWS account may trust the EC2 service principal. IAM administrators must grant the `iam:PassRole` permission to `IAM principals` to allow those principals in the account to launch instances with specific roles. A Cluster Operator attempts to create an `EC2NodeClass` with an IAM role not intended for Karpenter

Reviewer: ```suggestion
**Threat:** Many IAM roles in an AWS account may trust the EC2 service principal. IAM administrators must grant the `iam:PassRole` permission to `IAM principals` to allow those principals in the account to launch instances with specific roles. A Cluster Operator attempts to create an `EC2NodeClass` with an IAM role not intended for Karpenter.
```

> **Mitigation:** Cluster Operators must enumerate the roles in the resource section of the IAM policy granted to the Karpenter role for the `iam:PassRole` action. Karpenter will fail to generate an instance profile if role that is specified in the `spec.role` section of the `EC2NodeClass` is not enumerated in the `iam:PassRole` permission.

Reviewer: ```suggestion
**Mitigation:** Cluster Operators must enumerate the roles in the resource section of the IAM policy granted to the Karpenter role for the `iam:PassRole` action. Karpenter will fail to generate an instance profile if a role that is specified in the `spec.role` section of the `EC2NodeClass` is not enumerated in the `iam:PassRole` permission.
```

#### Execution of Arbitrary UserData

> **Threat:** Karpenter allows users to specify custom UserData as part of the `EC2NodeClass` resource. This UserData may take two shapes: a MIME multi-part archive or a NodeConfig yaml object. Karpenter does not perform any validation on this input, other than parsing the MIME multi-part archive if a header is detected. This custom UserData is then merged with UserData generated by Karpenter and is included in the Launch Template used to launch Karpenter managed nodes. Users with EDIT permissions on `EC2NodeClass` are able arbitrarily execute code on nodes launched by Karpenter.

Reviewer: Can we reference out to our userData section of our EC2NodeClass doc where we state that we support these formats? Also, this isn't true for all forms of AMIs so it may not be necessary mentioning the format at all here

> **Threat:** Karpenter allows users to specify custom UserData as part of the `EC2NodeClass` resource. This UserData may take two shapes: a MIME multi-part archive or a NodeConfig yaml object. Karpenter does not perform any validation on this input, other than parsing the MIME multi-part archive if a header is detected. This custom UserData is then merged with UserData generated by Karpenter and is included in the Launch Template used to launch Karpenter managed nodes. Users with EDIT permissions on `EC2NodeClass` are able arbitrarily execute code on nodes launched by Karpenter.

Reviewer: ```suggestion
**Threat:** Karpenter allows users to specify custom UserData as part of the `EC2NodeClass` resource. This UserData may take two shapes: a MIME multi-part archive or a NodeConfig yaml object. Karpenter does not perform any validation on this input, other than parsing the MIME multi-part archive if a header is detected. This custom UserData is then merged with UserData generated by Karpenter and is included in the Launch Template used to launch Karpenter managed nodes. Users with EDIT permissions on `EC2NodeClasses` can arbitrarily execute code on nodes launched by Karpenter.
```

> **Mitigation:** Access to `EC2NodeClass` is treated as a privileged permission. Access to the `EC2NodeClass` enables a user to not only specify custom UserData but also modify the IAM role used by the instance profile generated for the node, select the Node’s subnets and security groups, and other privileged settings. By default, only the Cluster Admin has access to this resource. If the Cluster Admin grants another user permission, this is the Cluster Admin’s responsibility. 

Reviewer: ```suggestion
**Mitigation:** Access to `EC2NodeClass` should be treated as a privileged permission. It enables a user to not only specify custom UserData but also modify the IAM role used by the instance profile generated for the node, select the Node’s subnets and security groups, and other privileged settings. By default, only the Cluster Admin should have access to this resource. 
```

> **Mitigation:** Access to `EC2NodeClass` is treated as a privileged permission. Access to the `EC2NodeClass` enables a user to not only specify custom UserData but also modify the IAM role used by the instance profile generated for the node, select the Node’s subnets and security groups, and other privileged settings. By default, only the Cluster Admin has access to this resource. If the Cluster Admin grants another user permission, this is the Cluster Admin’s responsibility. 

Reviewer: Can we also make a note that another mitigation that can be applied here is scoping down the permissions assigned to the nodeRole so that even if someone is able to execute arbitrary code, they can't do much in the AWS account?

### Denial of Service

> #### Repeated DescribeImages Call to the EC2 API

Reviewer: Why is this specific to DescribeImage? Isn't this any READ call that we support and that we would call based on the number of EC2NodeClasses/NodePools?

Reviewer: Can we generalize this so that it doesn't just apply to images?
